### PR TITLE
Support CTE

### PR DESCRIPTION
**Pull request by momosetkn**
Support CTE.
But not supported MySQL5.

Write it like this in Java: 

```java
var dcCte1 = new DepartmentCount_("dcCte1");
var dcCte2 = new DepartmentCount_("dcCte2");
var dcCte1Query =
    dsl.from(dCteInner1)
        .leftJoin(eCteInner1, on -> on.eq(dCteInner1.departmentId, eCteInner1.departmentId))
        .groupBy(dCteInner1.departmentId)
        .select(dCteInner1.departmentId, Expressions.count(eCteInner1.addressId));
var dcCte2Query =
    dsl.from(dCteInner2)
        .leftJoin(eCteInner2, on -> on.eq(dCteInner2.departmentId, eCteInner2.departmentId))
        .groupBy(dCteInner2.departmentId)
        .select(dCteInner2.departmentId, addOne(Expressions.count(eCteInner2.addressId)));
var query =
    dsl.with(dcCte1, dcCte1Query)
        .with(dcCte2, dcCte2Query)
        .from(e)
        .leftJoin(dcCte1, on -> on.eq(e.departmentId, dcCte1.departmentId))
        .leftJoin(dcCte2, on -> on.eq(e.departmentId, dcCte2.departmentId));
var list = query.fetch();
```

Write it like this in Kotlin: 

```kotlin
val dcCte1 = DepartmentCount_("dcCte1")
val dcCte2 = DepartmentCount_("dcCte2")
val dcCte1Query =
    dsl.from(dCteInner1)
        .leftJoin(eCteInner1) {
            eq(dCteInner1.departmentId, eCteInner1.departmentId)
        }
        .groupBy(dCteInner1.departmentId)
        .select(dCteInner1.departmentId, Expressions.count(eCteInner1.addressId))
val dcCte2Query =
    dsl.from(dCteInner2)
        .leftJoin(eCteInner2) {
            eq(dCteInner2.departmentId, eCteInner2.departmentId)
        }
        .groupBy(dCteInner2.departmentId)
        .select(dCteInner2.departmentId, addOne(Expressions.count(eCteInner2.addressId)))
val query =
    dsl.with(
        dcCte1 to dcCte1Query,
        dcCte2 to dcCte2Query
    )
        .from(e)
        .leftJoin(dcCte1) {
            eq(e.departmentId, dcCte1.departmentId)
        }
        .leftJoin(dcCte2) {
            eq(e.departmentId, dcCte2.departmentId)
        }
val list = query.fetch();
```

this example is generate bellow sql(formated)

```sql
with dcCte1(DEPARTMENT_ID, COUNT) as (select t3_.DEPARTMENT_ID, count(t4_.ADDRESS_ID)
                                      from DEPARTMENT t3_
                                               left outer join EMPLOYEE t4_ on (t3_.DEPARTMENT_ID = t4_.DEPARTMENT_ID)
                                      group by t3_.DEPARTMENT_ID),
     dcCte2(DEPARTMENT_ID, COUNT) as (select t3_.DEPARTMENT_ID, count(t4_.ADDRESS_ID) + 1
                                      from DEPARTMENT t3_
                                               left outer join EMPLOYEE t4_ on (t3_.DEPARTMENT_ID = t4_.DEPARTMENT_ID)
                                      group by t3_.DEPARTMENT_ID)
select t0_.EMPLOYEE_ID,
       t0_.EMPLOYEE_NO,
       t0_.EMPLOYEE_NAME,
       t0_.MANAGER_ID,
       t0_.HIREDATE,
       t0_.SALARY,
       t0_.DEPARTMENT_ID,
       t0_.ADDRESS_ID,
       t0_.VERSION
from EMPLOYEE t0_
         left outer join dcCte1 t1_ on (t0_.DEPARTMENT_ID = t1_.DEPARTMENT_ID)
         left outer join dcCte2 t2_ on (t0_.DEPARTMENT_ID = t2_.DEPARTMENT_ID);
```

### CODE REVIEW

File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/query/BuilderSupport.java [hunk 1]
```diff
@@ -62,6 +63,40 @@ public BuilderSupport(
...
+      buf.appendSql(") as (");
+      // cte body
+      SetOperationBuilder builder =
+          new SetOperationBuilder(config, setOperationContext, commenter, buf, aliasManager);
+      builder.build();
+      buf.appendSql("), ");
+    }
+    buf.cutBackSql(2);
+    buf.appendSql(" ");
+  }
```
Comment (momosetkn) [inline doma-core/src/main/java/org/seasar/doma/jdbc/criteria/query/BuilderSupport.java:98]: possible to common with the following code.
but desirable explicitly CTE column.
https://github.com/domaframework/doma/blob/add3be4dcdc8f4363b23fcbb79c6a8d5aa7093bc/doma-core/src/main/java/org/seasar/doma/jdbc/criteria/query/BuilderSupport.java#L100-L136

I think better A than B

A
```sql
with dcCte1(DEPARTMENT_ID, COUNT) as (select t3_.DEPARTMENT_ID, count(t4_.ADDRESS_ID)
                                      from DEPARTMENT t3_
                                               left outer join EMPLOYEE t4_ on (t3_.DEPARTMENT_ID = t4_.DEPARTMENT_ID)
                                      group by t3_.DEPARTMENT_ID),
```

B
```sql
with dcCte1 as (select t3_.DEPARTMENT_ID as DEPARTMENT_ID, count(t4_.ADDRESS_ID) as COUNT
                                      from DEPARTMENT t3_
                                               left outer join EMPLOYEE t4_ on (t3_.DEPARTMENT_ID = t4_.DEPARTMENT_ID)
                                      group by t3_.DEPARTMENT_ID),
```

File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/QueryDsl.java [hunk 2]
```diff
@@ -35,6 +37,294 @@ public QueryDsl(Config config) {
...
+            new WithContext(entityMetamodel2, subQuery2),
+            new WithContext(entityMetamodel3, subQuery3),
+            new WithContext(entityMetamodel4, subQuery4),
+            new WithContext(entityMetamodel5, subQuery5),
+            new WithContext(entityMetamodel6, subQuery6),
+            new WithContext(entityMetamodel7, subQuery7),
+            new WithContext(entityMetamodel8, subQuery8),
+            new WithContext(entityMetamodel9, subQuery9));
+    return with(withContexts);
+  }
```
Comment (nakamura-to) [inline doma-core/src/main/java/org/seasar/doma/jdbc/criteria/QueryDsl.java:322]: It seems that overloads with a large number of parameters are inconvenient to use. How about replacing them with a method chaining approach using 'with'?

For example, you could add the following code to `WithQueryDsl`:

```java
public WithQueryDsl with(EntityMetamodel<?> entityMetamodel, SetOperand<?> subQuery) { ... }
```

The usage code would look like this:

```java
var query = dsl
    .with(dcCte1, dcCte1Query)
    .with(dcCte2, dcCte2Query)
    .from(e)
```

File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/context/SelectContext.java [hunk 3]
```diff
@@ -44,6 +45,17 @@ public SelectContext(
...
+  }
+
+  public SelectContext(
+      EntityMetamodel<?> entityMetamodel,
+      Optional<SetOperationContext<?>> setOperationContextForSubQuery,
+      List<WithContext> withContexts) {
+    this.entityMetamodel = Objects.requireNonNull(entityMetamodel);
+    this.setOperationContextForSubQuery = Objects.requireNonNull(setOperationContextForSubQuery);
+    this.projection = new Projection.EntityMetamodels(entityMetamodel);
+    this.withContexts = withContexts;
```
Comment (nakamura-to) [inline doma-core/src/main/java/org/seasar/doma/jdbc/criteria/context/SelectContext.java:58]: Please ensure that the parameter of `withContexts` is not null.

**Review by nakamura-to — CHANGES_REQUESTED** [review]
@momosetkn Thank you for your contribution. I made some comments. I would appreciate it if you could address them.

File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/QueryDsl.java (hunk 2 above)
Comment (momosetkn) [reply doma-core/src/main/java/org/seasar/doma/jdbc/criteria/QueryDsl.java]: ok, I fixed!

Comment (nakamura-to) [reply doma-core/src/main/java/org/seasar/doma/jdbc/criteria/QueryDsl.java]: Thanks!

**Review by nakamura-to — APPROVED** [review]

**Pull request merged**